### DISCUSSION TITLE
Why do babel generate an extra function when transforming async function to generator?

### DISCUSSION THREAD

**Discussion by Austaras** [babel/babel#13902] [Q&A] [answered]
Why do babel transform following code
```js
async function bar() {}
```
to
```js
function bar() {
  return _bar.apply(this, arguments);
}

function _bar() {
  _bar = _asyncToGenerator(function* () {});
  return _bar.apply(this, arguments);
}
```
instead of something like
```js
function bar() {
  return _asyncToGenerator(function* () {}).apply(this, arguments);
}
```

**Comment by JLHwung** [comment, marked as answer]
They are transformed in this way so that `_asyncToGenerator` will be executed at most once.
Let's say you call the async function `bar`
```js
bar(1);
bar(2);
```

The first invocation `bar(1)` will evaluate the declared function `_bar` and then overwrite `_bar` with `_asyncToGenerator(function* () {});`. Any subsequent invocation `bar(2)`, etc. will be thus redirected to the cached `_bar`, the result of `_asyncToGenerator(function* () {});`.

**Reply by Austaras** [reply]
Could you also explain why [this](https://babeljs.io/repl#?browsers=chrome%2050&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=B4OgRgBAvBCGDOBPAdgYwgMwK5oC4EsB7ZCACgEoIBvAXwCg6AbAU1wkhgRXWzyJIrUaQA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.16.4&externalPlugins=&assumptions=%7B%7D) happens?

**Reply by JLHwung** [reply]
https://github.com/babel/babel/blob/473f145b27681a596815e6f8f50bc98f4fc48a7e/packages/babel-helper-wrap-function/src/index.js#L121-L127

The wrapper is retained for `let b = async function () {}` because when LHS is an identifier, the function expression in RHS will be executed by NamedEvaluation ([spec](https://tc39.es/ecma262/#sec-assignment-operators-runtime-semantics-evaluation)), in other words, we need to ensure that the new function (returned inside the wrapper) is named as `"b"`.

When LHS is a member expression, NamedEvaluation is not applied so we can safely omit the wrapper and output less code.
